ratbag/drivers/sinowealth.py

```diff
@@ -0,0 +1,318 @@
+        except KeyError:
+            logger.error(
+                f"Invalid report rate {obj.config & CONFIG_FLAGS_REPORT_RATE_MASK}"
+            )
+            report_rate = 0
+
+        xy_independent = bool(obj.config & CONFIG_FLAGS_XY_RESOLUTION)
+
+        def raw2dpi(raw: int) -> int:
+            # TODO: support different sensors. This is for PWM3360.
+            return (raw + 1) * 100
+
+        converted = [raw2dpi(r) for r in obj.dpis]
+        if xy_independent:
+            dpis = list(zip(converted[::2], converted[1::2]))
+        else:
+            dpis = [(x, x) for x in converted[:8]]
+
+        # Now create the Config object with all the data we have converted
+        # already
+        return cls(
+            dpi_disabled_slots=obj.dpi_disabled_slots,
+            report_rate=report_rate,
+            independent_xy_resolution=xy_independent,
+            dpis=dpis,
+        )
+
+
+@attr.s
+class Reply(object):
+    """
+    Object returned from :meth:`Query.run`. This object will have custom
+    attributes depending on the query's reply spec.
+    """
+
+    pass
+
+
+@attr.s
+class Query:
+    report_id: ReportID = attr.ib()
+    reply_report_id: ReportID = attr.ib()
+    cmd: Command = attr.ib()
+    query_spec: List[Spec] = attr.ib()
+    reply_spec: List[Spec] = attr.ib()
+
+    def run(self, rodent: ratbag.driver.Rodent) -> Reply:
+        # Use the long config if available, otherwise the short config
+        if (
+            self.reply_report_id == ReportID.CONFIG
+            and ReportID.CONFIG_LONG in rodent.report_ids["feature"]
+        ):
+            self.reply_report_id = ReportID.CONFIG_LONG
+
+        qspec = [
+            Spec("B", "report_id"),
+            Spec("B", "cmd"),
+        ] + self.query_spec
+
+        query = Parser.from_object(self, qspec, pad_to=self.report_id.size)
+        rodent.hid_set_feature(self.report_id, bytes(query))
+
+        reply_data = rodent.hid_get_feature(self.reply_report_id)
+        # Special handling for config queries:
+        if self.reply_report_id in (ReportID.CONFIG, ReportID.CONFIG_LONG):
+            if len(reply_data) != self.reply_report_id.size:
+                raise ratbag.driver.ProtocolError.from_rodent(
+                    rodent, f"Unexpected reply data size {len(reply_data)}"
+                )
+        # Real error check
+        elif len(reply_data) != len(query):
+            raise ratbag.driver.ProtocolError.from_rodent(
+                rodent, f"Unexpected reply data size {len(reply_data)}"
+            )
+
+        if self.reply_report_id != reply_data[0]:
+            raise ratbag.driver.ProtocolError.from_rodent(
+                rodent,
+                f"Got reply for a different command {reply_data[1]} instead of {self.reply_report_id}",
+            )
+
+        reply = Reply()
+        Parser.to_object(reply_data, self.reply_spec, obj=reply)
+        self.parse_reply(reply)
+        return reply
```

Comment:
I've played with this a bit in the `hidpp20` driver and found a nice solution for debugging, see 5d6beec22de47188590122463d48be6a94db85e8.

You can pass a class name (or type) into `to_object` and it'll create an `attr` class with that name which has a reasonable default `__str__`. This may be interesting here, see the first hunk of the linked commit.

Reply:
I don't understand, would something change this way?
Is this related in any way with `Cannot access member "version" for type "Reply"    Member "version" is unknown` warnings? (these are Pyright warnings, but mypy can do them too if you use `--strict` argument).

Reply:
the main advantage is that you get a better debugging output since it's no longer a generic `Reply` object but a named object (with `attr` attributes).

I'm not sure how to get rid of this warning though because all fields on the object are dynamic, so mypy/pyright can't check anything here. I'm currently experimenting with [this approach](https://github.com/whot/ratbag-python/commit/049fd7deebcb62f6df6377479fa13308b5a42594) in the hidpp20 driver for LEDs. Could be used for query/reply handling this way too and would help with those warnings but I'm not sure it's worth it because it requires some duplication. Just as a FYI.